2.MLP_Multi_Classifier/MLP_utils/utils.py

```diff
@@ -0,0 +1,938 @@
+        when return_info=True
+        int: returns metric specified in metric arg
+        when return_info=False
+
+    Raises
+    ------
+    optuna.exceptions.TrialPruned
+        exception is raised if/when a trial is pruned due to poor intermediate values
+    Exception
+        raised when metric value is nor 'accuracy' or 'loss'
+    """
+
+    # calling model function
+    model = build_model_custom(trial, in_features, out_features, params)
+
+    # param dictionary for optimization
+    optimization_params = {
+        "learning_rate": trial.suggest_float(
+            "learning_rate", params.LEARNING_RATE_MIN, params.LEARNING_RATE_MAX
+        ),
+        "optimizer": trial.suggest_categorical(
+            "optimizer", ast.literal_eval(params.OPTIMIZER_LIST)
+        ),
+    }
+
+    # param optimizer pick
+    optimizer = getattr(optim, optimization_params["optimizer"])(
+        model.parameters(), lr=optimization_params["learning_rate"]
+    )
+    # loss function
+
+    # for binary model use different for multi-class
+    criterion = nn.CrossEntropyLoss()
+
+    # send model to device(cuda)
+    model = model.to(params.DEVICE)
+    criterion = criterion.to(params.DEVICE)
+
+    # train set accuracy and loss
+    train_acc = []
+    train_loss = []
+
+    # validation set accuracy and loss
+    valid_acc = []
+    valid_loss = []
+
+    # total number of data to pass through
+    total_step = len(train_loader)
+    total_step_val = len(valid_loader)
+
+    for epoch in range(params.OPTIM_EPOCHS):
+
+        (
+            model,
+            optimizer,
+            criterion,
+            train_acc,
+            train_loss,
+            valid_acc,
+            valid_loss,
+            correct,
+            total,
+            correct_v,
+            total_v,
+        ) = train_n_validate(
+            model,
+            optimizer,
+            criterion,
+            train_acc,
+            train_loss,
+            valid_acc,
+            valid_loss,
+            total_step,
+            total_step_val,
+            params,
+            train_loader,
+            valid_loader,
+        )
+        if metric == "accuracy":
+            trial.report(np.mean(valid_acc), epoch)
+        elif metric == "loss":
+            trial.report(np.mean(valid_loss), epoch)
+        else:
+            raise Exception("metric not defined as accuracy or loss")
+
+        # Handle pruning based on the intermediate value
+        if trial.should_prune():
+            raise optuna.exceptions.TrialPruned()
+
+    # I want information returned but only 1 metric required for the optimize function called by study.optimize
+    # with out this conditional statement the optimization will fail
+    if return_info == True:
+        print(f"Validation Accuracy: {np.mean(valid_acc)}")
+        print(f"Validation Loss: {np.mean(valid_loss)}")
+        print(f"Training Accuracy: {np.mean(train_acc)}")
+        print(f"Training Loss: {np.mean(train_loss)}")
```

Comment:
Seeing that this if branch does not make a return and the others below do, consider revising this to keep consistent with returns for the function (either returning something or not for all branches / cases). Doing this would help improve readability and expectations within the function. Please use your best judgement on this comment; this is more of a recommendation than strict rule.